internal/dag/builder.go

```diff
@@ -512,6 +512,91 @@ func (b *Builder) computeHTTPProxy(proxy *projcontour.HTTPProxy) {
 	}
 }
 
+// expandPrefixMatches adds new Routes to account for the difference
+// between prefix replacement when matching on '/foo' and '/foo/'.
+//
+// The table below shows the behavior of Envoy prefix rewrite. If we
+// match on only `/foo` or `/foo/`, then the unwanted rewrites marked
+// with X can result. This means that we need to generate separate
+// prefix matches (and replacements) for the two cases.
+//
+// | Matching Prefix | Replacement | Client Path | Rewritten Path |
+// |-----------------|-------------|-------------|----------------|
+// | `/foo`          | `/bar`      | `/foosball` |   `/barsball`  |
+// | `/foo/`         | `/bar`      | `/foo/type` | X `/bartype`   |
+// | `/foo`          | `/bar/`     | `/foosball` | X `/bar/sball` |
+// | `/foo/`         | `/bar/`     | `/foo/type` |   `/bar/type`  |
+func expandPrefixMatches(routes []*Route) []*Route {
+	prefixedRoutes := map[string][]*Route{}
+
+	expandedRoutes := []*Route{}
+
+	// First, we group the Routes by their slash-consistent prefix match condition.
+	for _, r := range routes {
+		routingPrefix := r.PathCondition.(*PrefixCondition).Prefix
```

Comment:
is this safe to do? PathCondition could be a *RegexCondition

Reply:
Fixed.

Reply:
ta